### PR TITLE
[Core: LorisForm] Add type declarations to function signatures

### PR DESCRIPTION
**Pull request by johnsaigle**
See #3878 for background.

### CODE REVIEW

File: php/libraries/LorisForm.class.inc [hunk 1]
```diff
@@ -1187,7 +1228,7 @@ class LorisForm
      *
      * @return bool true if data was submitted by the user to the form.
      */
-    function isSubmitted()
+    function isSubmitted(): bool
```
Comment (PapillonMcGill) [inline php/libraries/LorisForm.class.inc]: skipped validate() on line 1240

File: php/libraries/LorisForm.class.inc [hunk 2]
```diff
@@ -2209,7 +2260,7 @@ class LorisForm
      *
      * @return string the error message of the element
      */
-    function getElementError($el)
+    function getElementError(string $el): string
```
Comment (PapillonMcGill) [inline php/libraries/LorisForm.class.inc]:
```suggestion
    function getElementError(string $el): ?string 
```
migth be void

File: php/libraries/LorisForm.class.inc [hunk 3]
```diff
@@ -2293,7 +2344,7 @@ class LorisFormFileElement extends LorisFormElement
      * @param array $el The element from $this->form which
      *                  we need an object wrapper around.
      */
-    function __construct($el)
+    function __construct(array $el)
```
Comment (PapillonMcGill) [inline php/libraries/LorisForm.class.inc]:
```suggestion
    function __construct(array $el): void
```
adjust doc block accordingly

File: php/libraries/LorisForm.class.inc [hunk 4]
```diff
@@ -2318,8 +2369,10 @@ class LorisFormFileElement extends LorisFormElement
      *
      * @return void
      */
-    function moveUploadedFile($destinationDir, $destinationFileName)
-    {
+    function moveUploadedFile(
+        string $destinationDir,
+        string $destinationFileName
+    ): void {
```
Comment (PapillonMcGill) [inline php/libraries/LorisForm.class.inc]: skipped getValue() on line 2389

File: php/libraries/LorisForm.class.inc (hunk 2 above)
Comment (johnsaigle) [reply php/libraries/LorisForm.class.inc]: I'll change the void return to return an empty string. Just FYI `?string` means null or string and doesn't cover void returns.

Comment (PapillonMcGill) [reply php/libraries/LorisForm.class.inc]: @johnsaigle Thanks for the precision about `?`

File: php/libraries/LorisForm.class.inc [hunk 5]
```diff
@@ -1187,14 +1228,14 @@ class LorisForm
...
     {
         return (!empty($_FILES) || !empty($_POST));
     }
     /**
      * Reimplements the HTML_QuickForm validate API.
      *
-     * @return true if all registered rules passed
+     * @return bool Whether all registed rules are passed.
      */
     function validate()
```
Comment (PapillonMcGill) [inline php/libraries/LorisForm.class.inc]:
```suggestion
    function validate(): bool
```

File: modules/user_accounts/php/edit_user.class.inc [hunk 6]
```diff
@@ -598,7 +598,7 @@ class Edit_User extends \NDB_Form
                 'UserID_Group',
                 'User name',
                 $this->_GUIDelimiter,
-                false
+                array()
```
Comment (HenriRabalais) [inline modules/user_accounts/php/edit_user.class.inc]: If the new declaration of `addGroup` will default the 5th parameter to an empty array, why include an empty here?

File: php/libraries/NDB_Page.class.inc [hunk 7]
```diff
@@ -177,8 +175,13 @@ class NDB_Page implements RequestHandlerInterface
...
      * @return void
      */
-    function addSelect($name, $label, $options, $optional=array())
-    {
+    function addSelect(
+        $name,
+        $label,
+        array $options = array(),
+        array $optional=array()
+    ) {
```
Comment (HenriRabalais) [inline php/libraries/NDB_Page.class.inc]: `: void`?

File: php/libraries/NDB_Page.class.inc [hunk 8]
```diff
@@ -434,8 +437,14 @@ class NDB_Page implements RequestHandlerInterface
...
      */
-    function addGroup($elements, $name, $label, $separator, $appendName=null)
-    {
+    function addGroup(
+        $elements,
+        $name = '',
+        $label = '',
+        $separator = '',
+        $appendName = array()
+    ) {
```
Comment (HenriRabalais) [inline php/libraries/NDB_Page.class.inc]: `: void`?

File: php/libraries/NDB_Page.class.inc [hunk 9]
```diff
@@ -566,11 +574,11 @@ class NDB_Page implements RequestHandlerInterface
      *
      * @param string $field   The fieldname for this radio button
      * @param string $label   The label to attach to this radio button
-     * @param string $options Options to pass to LorisForm
+     * @param array  $options Options to pass to LorisForm
      *
      * @return array representing radio button
      */
-    function createRadio($field, $label, $options=null)
+    function createRadio($field, $label, $options = array())
```
Comment (HenriRabalais) [inline php/libraries/NDB_Page.class.inc]: `: array`?

File: php/libraries/NDB_Page.class.inc [hunk 10]
```diff
@@ -545,12 +553,12 @@ class NDB_Page implements RequestHandlerInterface
...
      * @param string $field   The fieldname for this checkbox
      * @param string $label   The label to attach to this checkbox
-     * @param string $options Options to pass to HTML_QuickForm
+     * @param array  $options Options to pass to HTML_QuickForm
      * @param string $closer  ?????
      *
      * @return array representing checkbox
      */
-    function createCheckbox($field, $label, $options=null, $closer='</label>')
+    function createCheckbox($field, $label, $options = array(), $closer='</label>')
```
Comment (HenriRabalais) [inline php/libraries/NDB_Page.class.inc]: `: array`?

File: modules/user_accounts/php/edit_user.class.inc (hunk 6 above)
Comment (johnsaigle) [reply modules/user_accounts/php/edit_user.class.inc]: I changed the calling code to fix some bugs that came up during testing, but you're right they now don't need to be added. I'll remove the 5th param call.

**Comment by johnsaigle** [comment]
The default parameter values of NDB_Page  are incompatible with those required for LorisForm. As a result, they must be updated together. Since NDB_Page return type declarations are being delayed until LORIS 22, this PR must also wait.

This PR will be closed for now and re-opened after the 21 release.

**Pull request closed without merging**